plugins/hive-plugin/src/main/java/edu/berkeley/ground/plugins/hive/GroundReadWriteNodeResource.java

```diff
@@ -0,0 +1,116 @@
+            String jsonString = mapper.writeValueAsString(nodeVersion);
+            String uri = PluginUtil.groundServerAddress + "nodes/versions";
+            PostMethod post = new PostMethod(uri);
+            post.setRequestEntity(PluginUtil.createRequestEntity(jsonString));
+            String response = PluginUtil.execute(post);
+            return constructNodeVersion(response);
+        } catch (IOException e) {
+            throw new GroundException(e);
+        }
+    }
+
+    public NodeVersion getNodeVersion(long nodeVersionId) throws GroundException {
+        GetMethod get = new GetMethod(PluginUtil.groundServerAddress + "nodes/versions/" + nodeVersionId);
+        try {
+            String response = PluginUtil.execute(get);
+            return this.constructNodeVersion(response);
+        } catch (IOException e) {
+            throw new GroundException(e);
+        }
+    }
+
+    Node getNode(String dbName) throws GroundException {
+        HttpMethod get = new GetMethod(PluginUtil.groundServerAddress + "nodes/" + dbName);
+        try {
+            String response = PluginUtil.execute(get);
+            if (response != null) {
+                return this.constructNode(response);
+            }
+            return null;
+        } catch (IOException e) {
+            throw new GroundException(e);
+        }
+    }
+
+    public List<Long> getTransitiveClosure(Long nodeVersionId) throws GroundException {
+        GetMethod get = new GetMethod(PluginUtil.groundServerAddress + "nodes/closure/" + nodeVersionId);
+        try {
+            return (List<Long>) PluginUtil.getVersionList(get);
+        } catch (IOException e) {
+            throw new GroundException(e);
+        }
+    }
+
+    public List<Long> getAdjacentNodes(Long prevVersionId, String edgeName) throws GroundException {
+        GetMethod get = new GetMethod(PluginUtil.groundServerAddress + "adjacent/" + prevVersionId + "/" + edgeName);
+        try {
+            return (List<Long>) PluginUtil.getVersionList(get);
+        } catch (IOException e) {
+            throw new GroundException(e);
+        }
+    }
+
+    // create a node using Tag
+    Node createNode(String name, Map<String, Tag> tagMap) throws GroundException {
+        try {
+            String encodedUri = PluginUtil.groundServerAddress + "nodes/" + URLEncoder.encode(name, "UTF-8");
+            PostMethod post = new PostMethod(encodedUri);
+            ObjectMapper mapper = new ObjectMapper();
+            String jsonString = mapper.writeValueAsString(tagMap);
+            post.setRequestEntity(PluginUtil.createRequestEntity(jsonString));
+            String response = PluginUtil.execute(post);
+            return this.constructNode(response);
+        } catch (IOException e) {
+            throw new GroundException(e);
+        }
+
+    }
+
+    // helper methods for Node and NodeVersion
+    private Node constructNode(String response) throws GroundException {
+        if (response == null) {
```

Comment:
In general, what do you suggest for this? I was originally using `Optional`s for this as is, from my understanding, canonical in Scala, but upon further research / reading, it seems like the creators of the Java `Optional` vehemently oppose that sort of usage.